File: android/app/src/main/java/dk/betterw4/android/feature/review/ReviewPromptStore.kt
package dk.betterw4.android.feature.review

import android.content.Context
import androidx.core.content.edit
import dagger.hilt.android.qualifiers.ApplicationContext
import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Persists launch timestamps and rating-prompt anti-spam flags.
 */
@Singleton
class ReviewPromptStore @Inject constructor(
    @ApplicationContext context: Context,
) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /**
     * Records one authenticated launch (caller ensures at most once per process).
     * Prunes launches older than [LAUNCH_WINDOW_DAYS].
     */
    fun recordLaunch(now: Instant = Instant.now()) {
        val cutoff = now.toEpochMilli() - LAUNCH_WINDOW_MS
        val pruned = launchTimestampsMs()
            .filter { it >= cutoff }
            .toMutableList()
        pruned.add(now.toEpochMilli())
        prefs.edit { putString(KEY_LAUNCH_TIMESTAMPS, pruned.joinToString(",")) }
    }

    fun launchCountInWindow(now: Instant = Instant.now()): Int {
        val cutoff = now.toEpochMilli() - LAUNCH_WINDOW_MS
        return launchTimestampsMs().count { it >= cutoff }
    }

    fun promptCount(): Int = prefs.getInt(KEY_PROMPT_COUNT, 0)

    fun lastPromptAtMillis(): Long = prefs.getLong(KEY_LAST_PROMPT_AT, 0L)

    fun completedPlayFlow(): Boolean = prefs.getBoolean(KEY_COMPLETED_PLAY, false)

    fun neverAsk(): Boolean = prefs.getBoolean(KEY_NEVER_ASK, false)

    fun markSoftPromptShown(now: Instant = Instant.now()) {
        prefs.edit {
            putInt(KEY_PROMPT_COUNT, promptCount() + 1)
            putLong(KEY_LAST_PROMPT_AT, now.toEpochMilli())
        }
    }

    fun markPlayFlowRequested() {
        prefs.edit {
            putBoolean(KEY_COMPLETED_PLAY, true)
            putBoolean(KEY_NEVER_ASK, true)
        }
    }

    fun markNeverAsk() {
        prefs.edit { putBoolean(KEY_NEVER_ASK, true) }
    }

    private fun launchTimestampsMs(): List<Long> {
        val raw = prefs.getString(KEY_LAUNCH_TIMESTAMPS, null).orEmpty()
        if (raw.isBlank()) return emptyList()
        return raw.split(',')
            .mapNotNull { it.trim().toLongOrNull() }
    }

    companion object {
        const val LAUNCH_WINDOW_DAYS = 14L
        private val LAUNCH_WINDOW_MS = LAUNCH_WINDOW_DAYS * 24L * 60L * 60L * 1000L
        private const val PREFS = "bl_review_prompt"
        private const val KEY_LAUNCH_TIMESTAMPS = "launch_timestamps"
        private const val KEY_PROMPT_COUNT = "prompt_count"
        private const val KEY_LAST_PROMPT_AT = "last_prompt_at"
        private const val KEY_COMPLETED_PLAY = "completed_play_flow"
        private const val KEY_NEVER_ASK = "never_ask"
    }
}

File: references/betterlectio/ios/BetterLectio/ReviewPromptStore.swift
import Foundation

/// Persists launch timestamps and rating-prompt anti-spam flags.
/// Mirrors Android `ReviewPromptStore`.
final class ReviewPromptStore: @unchecked Sendable {
    static let shared = ReviewPromptStore()

    static let launchWindowDays: Int = 14

    private let defaults: UserDefaults
    private let prefix = "bl_review_prompt."

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    /// Records one authenticated launch (caller ensures at most once per process).
    /// Prunes launches older than ``launchWindowDays``.
    func recordLaunch(now: Date = Date()) {
        let cutoff = now.addingTimeInterval(-Self.launchWindowSeconds)
        var pruned = launchTimestamps().filter { $0 >= cutoff }
        pruned.append(now)
        defaults.set(pruned.map(\.timeIntervalSince1970), forKey: key(Self.keyLaunchTimestamps))
    }

    func launchCountInWindow(now: Date = Date()) -> Int {
        let cutoff = now.addingTimeInterval(-Self.launchWindowSeconds)
        return launchTimestamps().filter { $0 >= cutoff }.count
    }

    func promptCount() -> Int {
        defaults.integer(forKey: key(Self.keyPromptCount))
    }

    func lastPromptAt() -> Date? {
        let ms = defaults.double(forKey: key(Self.keyLastPromptAt))
        guard ms > 0 else { return nil }
        return Date(timeIntervalSince1970: ms)
    }

    func completedStoreFlow() -> Bool {
        defaults.bool(forKey: key(Self.keyCompletedStore))
    }

    func neverAsk() -> Bool {
        defaults.bool(forKey: key(Self.keyNeverAsk))
    }

    func markSoftPromptShown(now: Date = Date()) {
        defaults.set(promptCount() + 1, forKey: key(Self.keyPromptCount))
        defaults.set(now.timeIntervalSince1970, forKey: key(Self.keyLastPromptAt))
    }

    func markStoreFlowRequested() {
        defaults.set(true, forKey: key(Self.keyCompletedStore))
        defaults.set(true, forKey: key(Self.keyNeverAsk))
    }

    func markNeverAsk() {
        defaults.set(true, forKey: key(Self.keyNeverAsk))
    }

    private func launchTimestamps() -> [Date] {
        guard let raw = defaults.array(forKey: key(Self.keyLaunchTimestamps)) as? [Double] else {
            return []
        }
        return raw.map { Date(timeIntervalSince1970: $0) }
    }

    private func key(_ name: String) -> String { prefix + name }

    private static let launchWindowSeconds: TimeInterval =
        TimeInterval(launchWindowDays) * 24 * 60 * 60

    private static let keyLaunchTimestamps = "launch_timestamps"
    private static let keyPromptCount = "prompt_count"
    private static let keyLastPromptAt = "last_prompt_at"
    private static let keyCompletedStore = "completed_store_flow"
    private static let keyNeverAsk = "never_ask"
}
